python/ray/includes/serialization.pxi

```diff
@@ -0,0 +1,337 @@
+        int64_t num_blocks = (right - left) // block_size
+        size_t chunk_size
+        int64_t prefix, suffix
+        int i
+
+    # Update right address
+    right = right - (num_blocks % num_threads) * block_size
+
+    # Now we divide these blocks between available threads. The remainder is
+    # handled separately.
+    chunk_size = (right - left) // num_threads
+    prefix = left - src
+    suffix = src + nbytes - right
+    # Now the data layout is
+    # | prefix | k * num_threads * block_size | suffix |.
+    #
+    # We have chunk_size = k * block_size, therefore the data layout is
+    # | prefix | num_threads * chunk_size | suffix |.
+    # Each thread gets a "chunk" of k blocks.
+
+    for i in prange(num_threads, nogil=True, num_threads=num_threads):
+        memcpy(dst + prefix + i * chunk_size, left + i * chunk_size,
+               chunk_size)
+        if i == 0:
+            memcpy(dst, src, prefix)
+        if i == num_threads - 1:
+            memcpy(dst + prefix + num_threads * chunk_size, right, suffix)
+
+
+# Buffer serialization format:
+# | i64 offset(protobuf) | i64 len(protobuf) | inband_data | pad(64) | \
+# | buffers | pad(8) | protobuf_data |
+def unpack_pickle5_buffers(Buffer buf):
+    cdef:
+        shared_ptr[CBuffer] _buffer = buf.buffer
+        const uint8_t *data = buf.buffer.get().Data()
+        size_t size = _buffer.get().Size()
+        CPythonObject python_object
+        CPythonBuffer *buffer_meta
+        c_string inband_data
+        int64_t protobuf_offset
+        int64_t protobuf_size
+        int32_t i
+        const uint8_t *buffers_segment
+    protobuf_offset = (<int64_t*>data)[0]
+    protobuf_size = (<int64_t*>data)[1]
+    if protobuf_size > INT32_MAX:
+        raise ValueError("Incorrect protobuf size. "
+                         "Maybe the buffer has been corrupted.")
+    if not python_object.ParseFromArray(data + protobuf_offset, <int32_t>protobuf_size):
+        raise ValueError("Protobuf object corrupted")
+    inband_data.append(<char*>(data + python_object.inband_data_offset()),
+                       <size_t>python_object.inband_data_size())
+    buffers_segment = data + python_object.buffers_offset()
+    pickled_buffers = []
+    # Now read buffer meta
+    for i in range(python_object.buffer_size()):
+        buffer_meta = <CPythonBuffer *>&python_object.buffer(i)
+        buffer = SubBuffer(buf)
+        buffer.buf = <void*>(buffers_segment + buffer_meta.address())
+        buffer.len = buffer_meta.length()
+        buffer.itemsize = buffer_meta.itemsize()
+        buffer.readonly = buffer_meta.readonly()
+        buffer.ndim = buffer_meta.ndim()
+        if buffer_meta.format().size() > 0:
+            buffer.format = buffer_meta.release_format().c_str()
+        else:
+            buffer.format = NULL
+        if buffer_meta.shape_size() > 0:
+            buffer.shape = <Py_ssize_t *>malloc(
```

Comment:
Is it guaranteed that python will call free on this later?

Reply:
yep, I also find out this problem. I need to release it manually.